### ISSUE TITLE
Support of expiring entries based on the initial write of the entry

### ISSUE THREAD

**Issue by jgoldhammer** [ben-manes/caffeine#70]
Maybe it is supported out of the box, but I haven´t discovered this option.

Background:
- I want to expire an entry in the cache after a certain time amount the entry was created initially in the cache. I don´t think that the other options expireAfterAccess and expireAfterWrite do not work here (https://github.com/ben-manes/caffeine/wiki/Eviction#time-based).
- the entry will be updated during this timeframe a lot, so expiringAfterWrite does not work for me...

Is this possible? Any chance to get this into the library?

Thanks,
Jens

**Comment by ben-manes** [comment]
Perhaps this is better handled in your code? There is a complexity budget and, when its only a little painful for a special case, I think the trade-off favors not adding.

You could avoid a write into the cache for an update by using a holder value. For example,
```java
Holder<V> holder = cache.asMap().putIfAbsent(key, new Holder<>(newValue));
if (holder != null) {
  holder.setValue(newValue);
}
```

This type of usage would make `expireAfterWrite` behave the way you want. There may be a few small quirks depending on other features.

**Comment by ben-manes** [comment]
This feature is implemented, but additional tests are needed before release. You would be able to perform this expiration by using,

```java
Caffeine.newBuilder().expiry(new Expiry<K, V> {
  public long expireAfterCreate(K key, V value, long currentTime) {
    // return expiration duration, e.g. 5 minutes
  }
  public long expireAfterUpdate(K key, V value, long currentTime, long currentDuration) {
    return currentDuration
  }
  public long expireAfterRead(K key, V value, long currentTime, long currentDuration) {
    return currentDuration;
  }
}).build()
```

**Comment by jgoldhammer** [comment]
Thanks, that is great.

**Comment by ben-manes** [comment]
Released 2.5.0
